Display full specfile(s) in stdout during cypress run

jennifer-shehane opened the pull request:
- close #4977 

### User facing changelog

Stdout printed during `cypress run` no longer truncates text such a filenames and instead wraps the text to fit within the width of the terminal.

### Additional details

The way tables are printed in stdout has been refactored.

##### Paths now insert newlines

- Previously, paths would extend beyond the width of the terminal or (even less ideal) would truncate the path with `...`, now we have a new method `formatPath` that when passed the width of the table's column it will be drawn within, will add newlines at the width so that the path will return. 

https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/lib/modes/run.js#L115

- Many places where we were just doing `console.log('path/to/thing')` have now been replaced with tables so we can properly wrap depending on the set column width.

https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/lib/modes/run.js#L859

##### Leftover width when < 100 width 

- Before, the algorithm was naive about throwing extra leftover sizing into the width of the *first* column, which would be confusing if you assigned the width of the column to `3` (1 left padding, 1 width, 1 right padding), it would sometimes print out at a width of `10` because if the other columns did not take up a width of `100` - they would throw the extra width into the first column. 
- Now the algorithm respects the width assigned and subtracts the width taken up by borders from the *widest* column. 

https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/lib/util/terminal.js#L161

##### Snapshots

- Many of the snapshot replacements have been updated in order to not disrupt the 100 width of the original table. When replacing text from a snapshot like `1028x468`, we now replace it with the smallest possible width (in this case a screenshot could be `0x0` at it's smallest) and add padding to equal the length of the original screenshot, so we replace the text like: `_.startPad('YxZ', p1.length)`. 

https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/test/support/helpers/e2e.coffee#L77

- Additionally, in one area replacing the text in the snapshot was *too late*. The width of the tables had already been calculated and drawn (with the newline returns), so replacing the text at that point would mess up all of the table alignments. This is when the stdout is printing the current working directory. In this case, we pass a `FAKE_CWD_PATH` env var during test execution - that when present during `run`, replaces the current working directory path *before* we insert the newlines. 

https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/lib/modes/run.js#L120

##### Bug in cli-table-3

I found a bug with cli-table-3 where padding passed to a cell is ignored if set to `0`. Opened an issue here: https://github.com/cli-table/cli-table3/issues/124

### How has the user experience changed?

#### Run Header

**Before**

<img width="828" alt="Screen Shot 2019-10-17 at 9 47 23 AM" src="https://user-images.githubusercontent.com/1271364/67014680-3f221d00-f0c3-11e9-8f37-6c6842a9193a.png">


**After**

<img width="825" alt="Screen Shot 2019-10-17 at 9 38 55 AM" src="https://user-images.githubusercontent.com/1271364/67014010-16e5ee80-f0c2-11e9-92f7-e932d0155efc.png">



### Run Results

**Before**

<img width="1236" alt="Screen Shot 2019-10-17 at 9 47 36 AM" src="https://user-images.githubusercontent.com/1271364/67014705-45b09480-f0c3-11e9-9506-1a6ba00fcfe1.png">


**After**

<img width="824" alt="Screen Shot 2019-10-17 at 9 38 30 AM" src="https://user-images.githubusercontent.com/1271364/67014030-206f5680-f0c2-11e9-8ece-1cac9cba4655.png">



### PR Tasks

- [x] Have tests been added/updated?
- [NA] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [NA] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?
- [NA] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

jennifer-shehane committed:
Begin inserting newlines for long specfiles in results table of stdout

jennifer-shehane committed:
clean up some decaff weirdness

bahmutov commented on packages/server/lib/util/terminal.js:
```diff
@@ -39,30 +25,26 @@ const getBordersLength = (left, right) => {
 const convertDecimalsToNumber = function (colWidths, cols) {
-  let diff
-  const sum = _.sum(colWidths)
+  // const sum = _.sum(colWidths)
```
a few commented out lines

jennifer-shehane replied:
Thanks for pointing this out. I removed this while experimenting - need to add back.

jennifer-shehane replied:
Actually, I might just remove this 😬

jennifer-shehane committed:
move newlines function into util

- call ‘formatPath’ on any area where the path could be too long and
needs to word wrap

jennifer-shehane committed:
Clean up some styles, padding that does nothing - get specname closer to symbol by 1 pixel

jennifer-shehane committed:
Commit new snapshot changes. Word wrapping, some spacing changes

jennifer-shehane committed:
remove more padding that does functionally nothing

jennifer-shehane committed:
Fix more of the extra unnecessary padding in 'Running' table

jennifer-shehane committed:
Update other CLI snapshots to match new spacing

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

jennifer-shehane committed:
Remove the error thrown when colWidths exceeds 100, I don't see this as necessary since we have snapshots covering the design of the columns.

jennifer-shehane committed:
Update some snapshots I missed

jennifer-shehane committed:
Add gaurd in newLine function in case there is no specfile name

jennifer-shehane committed:
fix more snapshots

jennifer-shehane committed:
Fix snapshot extra space

brian-mann reviewed (dismissed):
- We need to leave the truncation in place for displaying the number of found spec files - because using a `specPattern` would list out potentially dozens or even hundreds of tests - which is not what the user wants.
- Add margin-right space back in for displaying the Running spec

brian-mann committed:
forcibly enable colors in all e2e tests, and strip them back out of the snapshots

jennifer-shehane committed:
Fix colors that got lost in stdout

jennifer-shehane committed:
Update newlines function to use simpler ramda split + fix implementation to remove last newline

- write unit test for newlines function for easier refactoring in future

jennifer-shehane committed:
iterating on fixing the weird column issues

jennifer-shehane commented on packages/server/lib/util/newlines.js:
```diff
@@ -0,0 +1,15 @@
+const addNewlineAtEveryNChar = (str, n) => {
```
Added a test and updated function to ramda splitEvery

bahmutov commented:
by the way, small pure functions the `addNewlineAtEveryNChar` are perfect for data -driven testing already built into `snap-shot-it`, https://github.com/bahmutov/snap-shot-it#data-driven-testing

jennifer-shehane committed:
handle if str is undefined in newlines function

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

# Conflicts:
#	packages/server/__snapshots__/2_config_spec.coffee.js

jennifer-shehane committed:
Draw the tables at exact 100 width, taking into account padding and borders

- Calculate the actual width of the table column in order to add
newlines at that length
- Add newlines for the ‘Finished processing’ output of video file
- Add newlines for the ‘Screenshots’ output of screenshot paths

jennifer-shehane committed:
remove .only

jennifer-shehane committed:
Remove extra 1 char padding

jennifer-shehane committed:
Remove color changes (separate PR for this)

jennifer-shehane committed:
Replace project path with same num chars of X's

We were just replacing it with foo/bar/baz which would mess up the
alignment of the snapshots since it was removing chars

jennifer-shehane commented:
So, the snapshots are really weird now, where I can't tell if the columns are lining up when there is a projectPath part of the new truncation. 

I implemented this change, where I replace the tmpDir and projectPath with the same num of chars of `XXXXX`: https://github.com/cypress-io/cypress/pull/5120/commits/50bc51744c20fe4075c2183636b32b1da75e09d1

But, I just realized this is really naive because when run on a different system with a different relative directory it will change the num of chars. 

So, I'm not sure what we want to do.

jennifer-shehane commented:
~Ok, so @brian-mann 	is suggesting that instead of trying to strip out the current Directory path later, that we just change the current working directory when running our internal tests.~

~I should look at the run code for video compression when our `--fake-cwd-path` env var is set.~

~Pass env var for `--fake-cwd-path`, if set then override `process.stdout.write` so that it will replace the current working directory returned from `console.log`.~

Pass env var for `--fake-cwd-path`, if set then within the cli-table3 tool, we should strip out the current working directory and replace it with a fake dir before calculating the newlines. This needs to be done in the `formatPath` function of the `run.js`

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

# Conflicts:
#	packages/server/__snapshots__/1_async_timeouts_spec.coffee.js
#	packages/server/__snapshots__/5_screenshots_spec.coffee.js
#	packages/server/lib/modes/run.js
#	packages/server/test/support/helpers/e2e.coffee

jennifer-shehane committed:
If we are running with a CYPRESS_ENV of 'test', replace the cwd

- Replace the current working directory if we’re running withing
Cypress so that our snapshots are consistent when run on diff machines.
- Ideally, I would have liked to have had a special ‘—fake-cwd-path’
flag so that we can override this later, but I couldn’t figure how to
pass that.

jennifer-shehane committed:
Continue stripping the projectPath from snapshot output

- This is required because some messaging comes from outside of the
run.js file, so we still need to replace these paths in snapshot.

jennifer-shehane committed:
Update snapshots to have correctly replaced cwd

jennifer-shehane committed:
Update formatting of new Node Version path to properly return within our stdout tables

jennifer-shehane committed:
remove tests for method that does not exist

- update the drawing of table columns to match the new logic - don’t
need weird spaces and padding right anymore.
- update snapshot to match

jennifer-shehane committed:
Update snapshot for performance test that didn't previously run

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

jennifer-shehane committed:
return empty string if no name passed (really just for weird edgecases in our own tests)

jennifer-shehane commented:
Why 😭  😭  😭  😭  😭  😭  😭 

<img width="771" alt="Screen Shot 2019-10-10 at 5 42 38 PM" src="https://user-images.githubusercontent.com/1271364/66608794-68075700-eb85-11e9-90e7-5d0e0d9d8978.png">

jennifer-shehane committed:
Add instructions on how to run performance tests in server

jennifer-shehane committed:
Update video processing and screenshot tables so they no longer use weird negative padding hack with ' ' border

jennifer-shehane committed:
Update snapshots to reflect new table spacing

jennifer-shehane committed:
Another snapshot update missed

jennifer-shehane committed:
Accept 'FAKE_CWD_PATH' env var to replace it in the run

jennifer-shehane committed:
Pad paren time to factor in extra whitespace

jennifer-shehane commented:
Exactly same issue after updating the startPadding for the `(X seconds)`

<img width="807" alt="Screen Shot 2019-10-16 at 9 10 21 AM" src="https://user-images.githubusercontent.com/1271364/66922348-e2583100-eff4-11e9-9e01-70293009f2d5.png">

jennifer-shehane commented:
Found the culprit, not sure how I didn't notice the video names were being replaced before. https://github.com/cypress-io/cypress/blob/issue-4977-specfiles-in-stdout/packages/server/test/support/helpers/e2e.coffee#L100

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

# Conflicts:
#
packages/server/__snapshots__/1_commands_outside_of_test_spec.coffee.js
#	packages/server/__snapshots__/2_browser_path_spec.coffee.js
#	packages/server/__snapshots__/2_cookies_spec.coffee.js
#	packages/server/__snapshots__/3_plugins_spec.coffee.js
#	packages/server/__snapshots__/3_user_agent_spec.coffee.js
#	packages/server/__snapshots__/4_request_spec.coffee.js
#	packages/server/__snapshots__/5_screenshots_spec.coffee.js
#	packages/server/__snapshots__/5_stdout_spec.coffee.js
#	packages/server/__snapshots__/5_subdomain_spec.coffee.js
#	packages/server/__snapshots__/6_web_security_spec.coffee.js

jennifer-shehane committed:
Update snapshots to no longer have static 'acb123.mp4' for video names

- This was throwing off the table calculations.

jennifer-shehane committed:
remove stripping screenshot dimensions + replace padStart with padEnd

jennifer-shehane committed:
Update snapshots to instead user shorter form 'second' in order to account for shorter duration printing when it is 1 second

jennifer-shehane committed:
Add correct padding for screenshot dimensions

jennifer-shehane committed:
update perf spec

jennifer-shehane committed:
Update performance snapshot

jennifer-shehane committed:
remove unused 'stripAnsi' dep

jennifer-shehane committed:
Merge branch 'develop' into issue-4977-specfiles-in-stdout

# Conflicts:
#
packages/server/__snapshots__/1_commands_outside_of_test_spec.coffee.js
#	packages/server/__snapshots__/2_cookies_spec.coffee.js
#	packages/server/__snapshots__/3_plugins_spec.coffee.js
#	packages/server/__snapshots__/3_user_agent_spec.coffee.js
#	packages/server/__snapshots__/4_request_spec.coffee.js
#	packages/server/__snapshots__/5_subdomain_spec.coffee.js
#	packages/server/__snapshots__/5_task_not_registered_spec.coffee.js
#	packages/server/__snapshots__/6_uncaught_support_file_spec.coffee.js
#	packages/server/test/e2e/5_stdout_spec.coffee
#	packages/server/test/support/helpers/e2e.coffee

jennifer-shehane committed:
Fix merge conflict failure - didn't regenerate snapshot correctly last time.

jennifer-shehane committed:
commit size of electron screenshot (fix incoming in another PR to avoid this)

flotwig commented on packages/server/test/support/helpers/e2e.coffee:
```diff
@@ -84,25 +90,35 @@ normalizeStdout = (str, options = {}) ->
-  .replace(/\((\d+ minutes?,\s+)?\d+ seconds?\)/g, "(X seconds)")
+  ## duration='1589' -> duration='XXXX'
+  .replace(/(duration\=\')(\d+)(\')/g, replaceDurationFromReporter) 
+  ## (15 seconds) -> (XX seconds)
```
thanks for adding these much-needed comments

flotwig commented on packages/server/test/support/helpers/e2e.coffee:
```diff
@@ -57,19 +57,25 @@ replaceDurationSeconds = (str, p1, p2, p3, p4) ->
-  ## Node Version: 1.2.3
-
+  ## Node Version: 10.2.3
   return p1 + _.padEnd("X", p2.length, "X") + ' (/foo/bar/node)' + ' '.repeat(p3.length - 16)
```
doesn't this make the replacement text size depend on the underlying node version?

Node 12.0.0 will become Node XXXXXX, Node 12.11.0 will become XXXXXXX

probably not worth too much time to fix since there's only 1 or 2 tests that use custom Node, but it will come up whenever the Node version string on CI changes length.

jennifer-shehane replied:
Yes, I know 😞 I didn't want to make anymore changes. 😓 I'll fix it

flotwig replied:
i think `1_system_node_spec` will be the only snapshot that changes

flotwig commented on packages/server/lib/util/terminal.js:
```diff
@@ -151,13 +128,11 @@ const wrapBordersInGray = (chars) => {
 const table = function (options = {}) {
-  const { colWidths, type } = options
-
+  let { colWidths, type } = options
```
`type` can still be `const`, right?

flotwig commented on packages/server/test/support/helpers/e2e.coffee:
```diff
@@ -57,19 +57,25 @@ replaceDurationSeconds = (str, p1, p2, p3, p4) ->
+replaceParenTime = (str, p1) ->
+  ## could be (1 second) or (10 seconds)
+  ## need to account for shortest and longest
+  return _.padStart("(X second)", p1.length)
```
this makes all the snapshots say `(X second)`, which reads a little weird, in normal English it would be `(X seconds)` since X is a variable

jennifer-shehane replied:
Yeah, this is what I came up with.

Do you have a suggestion for writing this so that it replaces all variants below with a single replacement?

```
(2 seconds)
(1 second)
(10 seconds)
```

flotwig replied:
ah, i see what the issue is, you can't replace `(1 second)` with `(X seconds)` without messing other stuff up. i guess you could update the regex to grab some extra whitespace from the beginning of the string, but that's beginning to feel really hacky. just doing `(X second)` is probably the best way then

flotwig reviewed (dismissed):
New stdout looks awesome! I noticed some minor things you might want to fix but all in all this looks great.

jennifer-shehane committed:
Correctly handle padding of node version so we don't have to rewrite this later.

jennifer-shehane committed:
split out let & const assignments
